src/ray/stats/metric_exporter.cc

```diff
@@ -0,0 +1,102 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <future>
+
+#include "ray/stats/metric_exporter.h"
+
+namespace ray {
+namespace stats {
+
+template <>
+void MetricExporter::ExportToPoints(
+    const opencensus::stats::ViewData::DataMap<opencensus::stats::Distribution>
+        &view_data,
+    const std::string &metric_name, std::vector<std::string> &keys,
+    MetricPoints &points) {
+  // Return if no raw data found in view map.
+  if (view_data.size() == 0) {
+    return;
+  }
+  // NOTE(lingxuan.zlx): No sampling in histogram data, so all points all be filled in.
+  std::unordered_map<std::string, std::string> tags;
+  for (size_t i = 0; i < view_data.begin()->first.size(); ++i) {
+    tags[keys[i]] = view_data.begin()->first[i];
+  }
+  // Histogram metric will be append suffix with mean/max/min.
+  for (const auto &row : view_data) {
+    MetricPoint mean_point{.metric_name = metric_name + ".mean",
+                           .timestamp = current_sys_time_ms(),
+                           .value = static_cast<double>(row.second.mean()),
+                           .tags = tags};
+    MetricPoint max_point{.metric_name = metric_name + ".max",
+                          .timestamp = current_sys_time_ms(),
+                          .value = static_cast<double>(row.second.max()),
+                          .tags = tags};
+    MetricPoint min_point{.metric_name = metric_name + ".min",
+                          .timestamp = current_sys_time_ms(),
+                          .value = static_cast<double>(row.second.min()),
+                          .tags = tags};
+    points.push_back(std::move(mean_point));
+    points.push_back(std::move(max_point));
+    points.push_back(std::move(min_point));
+    RAY_LOG(DEBUG) << "Metric name " << metric_name
+                   << ", mean value : " << mean_point.value
+                   << " max value : " << max_point.value
+                   << "  min value : " << min_point.value;
+
+    if (points.size() >= kMaxBatchSize) {
+      RAY_LOG(DEBUG) << "Point size : " << points.size();
+      metric_exporter_client_->ReportMetrics(points);
```

Comment:
If `points.size()` is really big, don't we need to split the report to multiple `kMaxBatchSize`? Let's say we have 1001 points with `kMaxBatchSize = 100`. Does this mean we will report 1001 points at once?

Reply:
> If `points.size()` is really big, don't we need to split the report to multiple `kMaxBatchSize`? Let's say we have 1001 points with `kMaxBatchSize = 100`. Does this mean we will report 1001 points at once?

Since the points are inserted into the queue one by one, in theory, there will not be too much data written at once.